### DISCUSSION TITLE
Support Lemmy, Kbin and the Fediverse as a whole

### DISCUSSION THREAD

**Discussion by MysteriousOrb2000** [Docile-Alligator/Infinity-For-Reddit#1490] [Ideas]
### Checklist

- [X] I have used the search function for [open](https://github.com/Docile-Alligator/Infinity-For-Reddit/discussions/categories/ideas) **and** [closed](https://github.com/Docile-Alligator/Infinity-For-Reddit/discussions/categories/completed-ideas) ideas to see if someone else has already submitted the same request.
- [X] I will describe the feature with as much detail as possible.

### Description

With Reddit pricing the API to an unreasonable price, it's clear that they're trying to kill off third party APIs. So my proposed solution is to add support for the Fediverse platforms that behave like Reddit like Lemmy or Kbin. Currently, there's not many Android apps for those two, so Infinity supporting would help while showing new people to the Fediverse.

### Additional context

This seem mostly back end work with minor changes to the UI/UX. Most of the Fediverse aggregate micro blogs applications try to copy Reddit's experience, so it shouldn't require too much work.

**Comment by MrLuxuri** [comment]
I hope @Docile-Alligator and the other devs see this! Also others please do upvote this discussion for better reach!

**Comment by njmdietrich** [comment]
Seconded. I have really grown to love infinity for reddit and would love infinity for lemmy! Would happily donate some cash for it to happen.

**Reply by DraconicNEO** [reply]
Agreed, I would definitely like to see Infinity for Lemmy. Yeah Jerboa is a nice app for Lemmy but I feel like having an Infinity client would be even better.

**Reply by captainnapalm** [reply]
I've grown so accustomed to how Infinity handles closing images/videos by swiping up that I do it in every app and am disappointed it doesn't work. I would love Infinity for Lemmy.

**Reply by DraconicNEO** [reply]
> I've grown so accustomed to how Infinity handles closing images/videos by swiping up that I do it in every app and am disappointed it doesn't work. I would love Infinity for Lemmy.

Seems like someone [made a fork to implement it](https://codeberg.org/Bazsalanszky/Eternity).

**Comment by andscape-dev** [comment]
This would be awesome indeed. It would probably require quite a bit of work though. Just mapping Lemmy's data models onto Reddit's is not a small task, let alone handling the differences in features and adapting the UI to them. Also, Lemmy and Kbin have different client APIs, so they would have to be integrated separately. There's a lot of developer time that would have to go into this.

**Reply by XGit3000** [reply]
A small price to pay compared to $20 million per year

**Reply by mdecourcy** [reply]
There's a Reddit -> Lemmy translation layer in development. Most recent update is [here](https://www.reddit.com/r/apihackathon/comments/142a236/redreader_browsing_fedditdecnetzpolitik_via_a/). Doesn't seem to have a git repo up yet.

**Reply by andscape-dev** [reply]
@mdecourcy that's a cool idea as a temporary replacement. I would love to see their repo. Hopefully the translation is done client side so there isn't a need for another hosted service. Also, props to them: Lemmy does not have an OpenAPI spec published, so the API has to be reverse engineered from the code, which is a lot of boring work.

Eventually it would be better for the app to integrate directly with the Lemmy API, though, so it can make use of Lemmy-specific features...

**Reply by mdecourcy** [reply]
@andscape-dev repo is up, it'll need to be hosted. I'd be willing to host in the interim. https://github.com/derivator/tafkars

**Reply by andscape-dev** [reply]
@mdecourcy awesome, thanks for the link! Unfortunately I've never used Rust so I'm not sure how much I'll be able to contribute. Is there a chat room or something similar to discuss it?

**Comment by meisme-dev** [comment]
https://github.com/derivator/tafkars This is a piece of software that maps the Lemmy API to function like the Reddit API, already partially works with libreddit and RedReader

**Comment by ortyrell** [comment]
###  First lemmy post.  Curious to see what unfolds over the next few weeks.

Bear with me as I practice but it's seemingly intuitive.

[Practice link.](https://www.wikipedia.org)

![image](https://github.com/Docile-Alligator/Infinity-For-Reddit/assets/136193800/dd40910b-9891-4fcf-85e9-b02f576c247a)

[Qwertyuiop.docx](https://github.com/Docile-Alligator/Infinity-For-Reddit/files/11713188/Qwertyuiop.docx)

**Reply by ortyrell** [reply]
Well, there it is.

**Reply by ortyrell** [reply]
Sort of.  Just figured out this is github.

**Comment by adriandelgg** [comment]
++1

**Comment by KSPAtlas** [comment]
Looking at the code, we would need to do roughly these things:
- Implement a way to select lemmy instance before the API gets initialized
 - Add a LemmyAPI.java
- Change the code that interfaces with the API to accept the lemmy API
- Remove any redundant code/features

**Comment by DraconicNEO** [comment]
This is an awesome idea, I hope that it gets implemented into Infinity. This client is honestly awesome and it would be even more awesome to be able to use it as a Lemmy or even Kbin client, especially since Kbin doesn't have its own app yet.

Though I kind of feel like we should separate the Reddit from kbin and Lemmy and have it be a separate version that does fediverse stuff. It would help because for people who are still going to use Reddit (either with their API key or using one extracted from the official app) probably wouldn't want it to be in the same app that they use for their fediverse stuff, at least I wouldn't.

**Comment by Darin755** [comment]
Honestly it would be nice to have a toggle between reddit and Lemmy. That probably wouldn't work with the reddit api changes but it would help people discover lemmy

**Reply by DraconicNEO** [reply]
Well it could be possible to add a section for people to add their own API key and client data to use Reddit in the future, it is technically against their terms of service. On its own that wouldn't really stop it from happening but since the dev wants to continue relations with Reddit, and said that he was going to try making Infinity paid exclusive on a monthly basis in order to work with the new API changes it obviously can't happen from him.

**Reply by aman207** [reply]
> Well it could be possible to add a section for people to add their own API key and client data to use Reddit in the future, it is technically against their terms of service. 

Someone posted a solution on how to build Infinity with your own API key https://old.reddit.com/r/Infinity_For_Reddit/comments/14c2v5x/build_your_own_apk_with_your_personal_api_key_in/

**Comment by DraconicNEO** [comment]
> > Well it could be possible to add a section for people to add their own API key and client data to use Reddit in the future, it is technically against their terms of service.
> 
> Someone posted a solution on how to build Infinity with your own API key https://old.reddit.com/r/Infinity_For_Reddit/comments/14c2v5x/build_your_own_apk_with_your_personal_api_key_in/



> > Well it could be possible to add a section for people to add their own API key and client data to use Reddit in the future, it is technically against their terms of service.
> 
> Someone posted a solution on how to build Infinity with your own API key https://old.reddit.com/r/Infinity_For_Reddit/comments/14c2v5x/build_your_own_apk_with_your_personal_api_key_in/

I think if one were to make a fork of Infinity that supports Federated services and the ability to switch back to using Reddit in the same app, they'd probably want to use a solution for the client data that isn't hard-coded. As in they'd want to allow people to input their own data in the settings.

Of course this is a discussion for forks, since like I said before docile alligator wants to continue relations with Reddit and [make Infinity subscription based](https://www.reddit.com/r/Infinity_For_Reddit/comments/147bhsg/the_future_of_infinity/) so he's obviously not going to do this himself (since that would definitely end relations with Reddit by breaking terms of service).

**Comment by JoshuaKimsey** [comment]
Thought I should post this here for others to see as well, but [Bazsalanszky](https://lemmy.toldi.eu/u/bazsalanszky) has made a fork of Infinity for Lemmy that is now up and running. Still some rough edges, but it's definitely the best work towards it I've seen thus far (and also now makes it one of the best apps for Lemmy now too). You can see the repository here on [Codeberg](https://codeberg.org/Bazsalanszky/Infinity-For-Lemmy). There's also a dedicated [community](https://lemdro.id/c/infinityforlemmy) for the app's development on Lemmy as well. 

It really is a beautiful sight seeing Infinity work (mostly) flawlessly with Lemmy.

**Reply by carlocastoldi** [reply]
I'm sorry if this wasn't the right place to ask, but as far as you know, would it be possible to extend Infinity/Eternity to aggregate posts both from Reddit and Lemmy?
